File: python/ecb/ecb_function.py
from __future__ import annotations

import io
import pandas as pd
import requests


def ecb_api_function(
    dataset: str,
    series_key: str,
    base_url: str = "https://data-api.ecb.europa.eu/service/data",
) -> pd.DataFrame:
    """
    Descargar datos del BCE (ECB Data) en CSV (csvdata) y devolver DataFrame.

    Parameters
    ----------
    dataset : str
        Identificador del dataset (p. ej., "BSI").
    series_key : str
        Clave completa de la serie (dimensiones concatenadas con '.').
    base_url : str
        Host del servicio de datos del BCE.
    """

    url = f"{base_url}/{dataset}/{series_key}"
    params = {"format": "csvdata"}
    resp = requests.get(url, params=params, headers={"Accept": "text/csv"}, timeout=120)
    resp.raise_for_status()

    text_stream = io.StringIO(resp.content.decode("utf-8"))
    df = pd.read_csv(text_stream)
    return df

File: R/ecb/ecb_function.R
# Función mínima BCE Data API (CSV)

# Inputs obligatorios:
# - dataset: identificador del dataset (p. ej., "BSI")
# - seriesKey: clave completa de la serie (dimensiones concatenadas con '.')
#
# Inputs opcionales:
# - base_url: host del servicio de datos del BCE
#
# Output:
# - df: data.frame con los datos descargados

ecb_api_function <- function(
  dataset,
  seriesKey,
  base_url = "https://data-api.ecb.europa.eu/service/data"
) {
  url <- paste0(base_url, "/", dataset, "/", seriesKey)
  res <- httr::GET(url, query = list(format = "csvdata"), httr::accept("text/csv"))
  if (httr::http_error(res)) stop(sprintf("ECB request failed [%s]", httr::status_code(res)))

  txt <- httr::content(res, as = "text", encoding = "UTF-8")
  df <- utils::read.csv(text = txt, stringsAsFactors = FALSE)
  return(df)
}


